src/couch_rate/src/couch_rate_ets.erl

```diff
@@ -0,0 +1,147 @@
+% Licensed under the Apache License, Version 2.0 (the "License"); you may not
+% use this file except in compliance with the License. You may obtain a copy of
+% the License at
+%
+%   http://www.apache.org/licenses/LICENSE-2.0
+%
+% Unless required by applicable law or agreed to in writing, software
+% distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+% WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+% License for the specific language governing permissions and limitations under
+% the License.
+
+-module(couch_rate_ets).
+
+-include("couch_rate.hrl").
+
+-export([
+    create_tables/0,
+    new/3,
+    from_map/3,
+    budget/1,
+    delay/1,
+    wait/1,
+    in/2,
+    success/2,
+    failure/1,
+    is_congestion/1,
+    min_latency/1,
+    format/1,
+    to_map/1,
+    module/1,
+    state/1,
+    table_name/1,
+    table_names/0
+]).
+
+
+-define(SHARDS_N, 16).
+
+-compile({parse_transform, lookup_pt}).
+
+%% The following directive would generate following code
```

Comment:
The difference is in efficiency. This line is executed at runtime https://github.com/apache/couchdb/blob/prototype/fdb-layer/src/fabric/src/fabric2_index.erl#L100. It does translation of list to atom and integer to list at runtime. With my design it is done once at compile time. At run time it is going to be just a clause match.

Reply:
That's pretty neat, though it does seem a bit complex as parse transforms are a rather advanced.

The impact of `list_to_atom/1` doesn't seem that bad - about 1-2 microseconds:

```
timer:tc(fun() -> lists:foreach(fun(_) -> list_to_atom("fabric2_index_" ++ integer_to_list(1)) end, lists:seq(1, 10000)) end).
{10960,ok}
```

And that's with the overhead of creating 10k item list.

Reply:
Let me know if you want me to remove `parse_transform`. I can easily replace it with hardcoded clauses:
```
table_name(0) -> couch_rate_ets_0;
table_name(1) -> couch_rate_ets_1;
table_name(2) -> couch_rate_ets_2;
table_name(3) -> couch_rate_ets_3;
table_name(4) -> couch_rate_ets_4;
table_name(5) -> couch_rate_ets_5;
table_name(6) -> couch_rate_ets_6;
table_name(7) -> couch_rate_ets_7;
table_name(8) -> couch_rate_ets_8;
table_name(9) -> couch_rate_ets_9;
table_name(10) -> couch_rate_ets_10;
table_name(11) -> couch_rate_ets_11;
table_name(12) -> couch_rate_ets_12;
table_name(13) -> couch_rate_ets_13;
table_name(14) -> couch_rate_ets_14;
table_name(15) -> couch_rate_ets_15;
table_name(Id) -> {error, Id}.
```

Reply:
Depending how often this code will be called. 1 microsecond difference every few seconds doesn't seem worth adding a parse transform or look up table for? If it is going to be called a 10000 times in a loop per single doc request, it would add a 10 millisecond to each request, then we would want to optimize it for sure.

So depending on that start with a simple one-line `list_to_atom()` to keep the code base smaller then later on, if we see a performance issue, add the parse transform? And we can also re-use it for other sharded tables once we added it. (We'd put in `couch` or `fabric` util function).

Reply:
I agree with @nickva on this. I would prefer we remove the transform unless it is a critical performance bottleneck. I would prefer adding itin as a separate commit to be used whereever we are adding sharded ets tables.

Reply:
Parse transform is removed in https://github.com/apache/couchdb/pull/2662/commits/d029086def2566455329c754d704d58b44be5744